feat: added a fix for npmjs now returns readme in version instead of global…

aremishevsky opened the pull request:
… metadata readme key

<!--

Before Pull Request check whether your commits follow this convention

https://github.com/verdaccio/verdaccio/blob/master/CONTRIBUTING.md#git-commit-guidelines

  * If your PR fix an issue don't forget to update the unit test and documentation in /docs folder
  * If your PR delivers a new feature, please, provide examples and why such feature should be considered.
  * Document your changes /docs
  * Add unit test
  * Follow the commit guidelines in order to get a quick approval

Pick one/multiple type, if none apply please suggest one, we might be included it by default

eg: bug / feature / documentation / unit test / build

-->
**Type:**

The following has been addressed in the PR:

*  There is a related issue?
*  Unit or Functional tests are included in the PR

**Description:**

<!-- Resolves #??? -->

juanpicado commented:
I just checked the code. I'll test it later with real data.

aremishevsky commented:
Hi @juanpicado 

As an example of dist-tags i've took a dist-tasg list from verdaccio package.
I've just added "next","dev" and "canary" tags in this list.
Please note it will take the readme of the first found tag's version readme.

juanpicado commented:
It seems to be fix https://github.com/npm/registry/issues/305

juanpicado commented on src/lib/storage-utils.js:
```diff
@@ -71,11 +71,34 @@ function cleanUpReadme(version: Version): Version {
+  const latestVersion = distTags['latest'] ? versions[distTags['latest']] || {} : {};
+  let readme = latestVersion.readme || pkg.readme || '';
+  // In case of empty readme - trying to get ANY readme in the following order: 'next','beta','alpha','test','dev','canary'
+  const readmeDistTagsPriority = [
```
If you catch here `readme` you do not need to iterate.
`let readme = latestVersion.readme || pkg.readme || '';`

Please return readme whether is different than an empty string. I'd add `|| ''` only as las resource.

Try to avoid iterations if are not need it.

juanpicado commented on src/lib/storage-utils.js:
```diff
@@ -71,11 +71,34 @@ function cleanUpReadme(version: Version): Version {
   return version;
 }
 
+function getLatestReadme(pkg: Package): string {
```
`getLatestReadme` function requires a unit test in a new file. I suggest `test/unit/latest.readme.spec.js`

aremishevsky commented:
Hi @juanpicado .
I've added some minor changes according to your review.
Could you please take a look?

juanpicado commented:
@aremishevsky I'll do it after work :) Thanks for your changes.  Please read this https://github.com/verdaccio/verdaccio/pull/677#issuecomment-387176345

aremishevsky commented:
@juanpicado 
I've changed the last commit.
However I can't change previous commits using rebase command. Not sure that I will not revert some of your changes.
If it's critical to have all commits with "feat: xxx" in messages - i can re-commit my changes in a new forked repo. If not - you can just merge it.
Just let me know

juanpicado commented:
![screen shot 2018-05-08 at 11 31 08 pm](https://user-images.githubusercontent.com/558752/39784294-354cbd98-5318-11e8-85ad-2ce4b0ecbfd5.png)

juanpicado commented:
I just noticed. Just above your code, there is a function called `cleanUpReadme` https://github.com/verdaccio/verdaccio/blob/master/src/lib/storage-utils.js#L66 which remove all readme from versions.

So, the result is gonna be always `undefined`. 

![screen shot 2018-05-08 at 11 31 08 pm](https://user-images.githubusercontent.com/558752/39784300-3cd7bfea-5318-11e8-9e79-30c995bd9c08.png)

So, I think

````
// In case of empty readme - trying to get ANY readme in the following order: 'next','beta','alpha','test','dev','canary'
  const readmeDistTagsPriority = [
    'next',
    'beta',
    'alpha',
    'test',
    'dev',
    'canary'];
  readmeDistTagsPriority.map(function(tag) {
    if(readme) {
      return readme;
    }
    const data = distTags[tag] ? versions[distTags[tag]] || {} : {};
    readme = _.trim(data.readme || readme);
  });
````

is a bit unecessary with the current approach, with local-storage we cannot afford save readme for each version otherwhise metadata will grow exponenxially, that's the main reason why is being removed.

About your commit would be something like this `"feat: return latest dis-tag readme whether readme is emtpy"

But you must remove the iteration, sorry I did not noticed that earlier.

juanpicado commented:
Some lint is failing 

```
/home/ubuntu/verdaccio/src/lib/storage-utils.js
  80:3  error  Expected space(s) after "if"  keyword-spacing
  92:5  error  Expected space(s) after "if"  keyword-spacing
````

aremishevsky commented:
Hi @juanpicado 

I've re-created a PR with a new single commit with a normal message: https://github.com/verdaccio/verdaccio/pull/687

As for iteration and cleanUpReadme calling - If you take a look closer - I've moved readme set before calling cleanup function - so it should be fine and it is needed.


![feat__added_a_fix_for_npmjs_now_returns_readme_in_version_instead_of_global _by_aremishevsky_ _pull_request__654_ _verdaccio_verdaccio](https://user-images.githubusercontent.com/32752187/39870045-8ff933b0-5468-11e8-8733-6eea3ebbe77b.jpg)


As for lint errors - i've fixed them in new PR. Thanks for an advise.
Please review this PR: https://github.com/verdaccio/verdaccio/pull/687

aremishevsky commented:
Hi @juanpicado .
Did you have a chance to take a look on https://github.com/verdaccio/verdaccio/pull/687 PR?

Thanks

juanpicado commented:
@aremishevsky give me some days, I'm un a hurry to finish Google Cloud plugin (before expiring my trial) and next week is the We Are Developers at Vienna 😓 ... I'll try to do it as soon as possible.

juanpicado commented:
Thanks, done :) thanks for your patient I'll merge #687 and close this one.

aremishevsky commented:
@juanpicado Thanks for review.
Could you please let me know once it would be published with a some next release version on npmjs?

juanpicado commented:
Just after my dinner 🍽 😉

juanpicado commented:
https://github.com/verdaccio/verdaccio/releases/tag/v3.0.0-beta.10
